src/storage/src/hummock/sstable/builder.rs

```diff
@@ -121,17 +150,40 @@ impl SstableBuilder {
 
         // TODO: refine me
         value.encode(&mut self.raw_value);
+        let mut extract_key = user_key(full_key);
         if let Some(table_id) = get_table_id(full_key) {
             if self.last_table_id != table_id {
                 self.table_ids.insert(table_id);
                 self.last_table_id = table_id;
             }
+
+            extract_key = match &mut self.filter_key_extractor {
+                Some(filter_key_extractor) => filter_key_extractor.extract(extract_key),
+
+                None => extract_key,
+            };
         }
+
+        // if let Some(table_id) = get_table_id(full_key) {
+        //     if self.last_table_id != table_id {
+        //         self.table_ids.insert(table_id);
+        //         self.last_table_id = table_id;
+        //     }
+        // }
         self.block_builder.add(full_key, self.raw_value.as_ref());
         self.raw_value.clear();
 
-        let user_key = user_key(full_key);
-        self.user_key_hashes.push(farmhash::fingerprint32(user_key));
+        // add bloom_filter check
+        // 1. not empty_key
+        // 2. transform key is not duplicate
+        if !extract_key.is_empty()
+            && (extract_key.len() > self.last_full_key.len()
+                || extract_key != &self.last_full_key[0..extract_key.len()])
```

Comment:
To reduce memory copy, we only use `self.last_full_key` when block_builder is empty, otherwise, we use last_full_key in block_builder.

Reply:
disscuss with wallace, for this case, we need to assign last_full_key every add call